src/engine/homeds/btree/btree.hpp

```diff
@@ -735,7 +757,182 @@ class Btree {
+                                                         LOCKTYPE_READ, nullptr);
+                        unlock_node(next_node, LOCKTYPE_READ);
+
+                        if (ret != btree_status_t::success) {
+                            LOGERROR("Cannot read sibling node for {}", m_next_left_node);
+                            resp.json["error"] = "Cannot read sibling node for {}" + m_next_left_node->to_string();
+                            return resp;
+                        }
+
+                        if (next_node->is_leaf() == false) {
+                            LOGERROR("next_node: {} is not leaf, not expected!", next_node->to_string());
+                            resp.json["error"] = "node: " + next_node->to_string() + " is not leaf, internal error. ";
+                            return resp;
+                        }
+
+                        m_next_left_node = next_node;
+                    }
+                }
+            }
+
+            //
+            // if we are here, it means it is a continue of next pagination,
+            // fall through
+            //
+        }
+
+        // now both range query and/or pagination starts here;
+        auto cur_node = m_next_left_node;
+        bool has_more = true;
+        for (int i = 0u; i < request.batch_size; ++i) {
+            // fill the response
+            std::string node_str = "leaf_" + std::to_string(i);
+            resp.json[node_str] = cur_node->to_string();
+
+            // get next node in chain;
+            if (cur_node->get_next_bnode() == empty_bnodeid) {
+                // no more leaf nodes, let's break out;
+                has_more = false;
+                break;
+            }
+
+            BtreeNodePtr next_node = nullptr;
+            auto ret =
+                read_and_lock_sibling(cur_node->get_next_bnode(), next_node, LOCKTYPE_READ, LOCKTYPE_READ, nullptr);
+            unlock_node(next_node, LOCKTYPE_READ);
+
+            HS_DBG_ASSERT_EQ(ret, btree_status_t::success);
+            if (ret != btree_status_t::success) {
+                LOGERROR("Cannot read sibling node for {}", cur_node);
+                resp.json["error"] = "Cannot read sibling node for {}" + cur_node->to_string();
+                return resp;
+            }
+
+            HS_DBG_ASSERT_EQ(next_node->is_leaf(), true);
+            cur_node = next_node;
+        }
+
+        if (has_more) {
+            // prepare for next batch
+            m_next_left_node = cur_node;
+            // remember where we are, this cursor also tells us whether this
+            // is a range query;
+            resp.json["next_cursor"] = m_next_left_node->get_node_id();
+            resp.json["has_more"] = "true";
+        } else {
+            m_next_left_node = nullptr;
+            resp.json["next_cursor"] = "no_more_nodes";
+            resp.json["has_more"] = "false";
+        }
```

Comment:
I think we are leaving the cur_node locked all the while during pagination. I think we should unlock the last node before returning.

Reply:
true, I am actually changing it so that no node should be in locked state after exiting this function.

Reply:
my original code won't hold any lock during pagination actually, because it always unlock the current, until next node is empty (which in fact is wrong); 

If any node is in locked state, the homestore shutdown will crash.